### PR TITLE
[ZCash] Upgrade ZCashOrchardStorage for shard tree purposes

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/42281

Part of https://github.com/brave/brave-core/pull/26477

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:206]: nit: prefer `const mojom::AccountIdPtr& account_id` here and below

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:601]: nit: I'd suggest to return "expected" values as base::ok()

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:595]: Having MAX() most likely requires indexes. Please add a TODO or an issue to research that

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:641]: `result.data = base::ToVector(resolve_cap.ColumnBlob(0))`

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.h
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.h:48]: Preferred way now is something like
base::SpanWriter(base::span(data)).Write(blob)
or
base::span(to).copy_from(blob)

File: components/brave_wallet/browser/zcash/zcash_shield_sync_service.cc
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_shield_sync_service.cc:211]: can this overflow?

File: components/brave_wallet/common/zcash_utils.h
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.h:141]: = 0

File: components/brave_wallet/common/zcash_utils.h
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.h:203]: = 0

File: components/brave_wallet/common/zcash_utils.h
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.h:255]: needs some default value

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:911]: Will ToVector work here?

File: components/brave_wallet/browser/zcash/zcash_shield_sync_service.cc
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_shield_sync_service.cc:223]: Added additional check, block height should not be less than 1,687,104.

File: components/brave_wallet/common/zcash_utils.cc
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.cc:113]: nit: static

File: components/brave_wallet/common/zcash_utils.cc
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.cc:124]: Does short version like `base::HexEncode(smth)` work?

**Review by supermassive — APPROVED** [review]
wallet core lgtm

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:198]: Is there an issue for this TODO?

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:969]: nit: the complexity of this function is pretty large. It might make sense to split this up a bit further in a refactor or other PR where you're touching this code later.

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:745]: same here WRT TODO

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (kdenhartog) [inline components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:167]: I think I asked this when we first added this, but can't remember where we landed. Are we encrypting this database?

**Review by kdenhartog — COMMENTED** [review]
nothing stands out as a blocker - few post merge comments, but nothing really needs to be changed so leaving approval for security review here.

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:167]: No we don't

File: components/brave_wallet/browser/zcash/zcash_orchard_storage.cc
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_orchard_storage.cc:198]: https://github.com/brave/brave-browser/issues/43529

**Pull request merged**
